### PR TITLE
ROI utils

### PR DESCRIPTION
**Pull request by szalpal**
#### Why we need this PR?
*Pick one*
- Adds some utility functions for Region of Intest (handy, for all kernels that require ROI)

**Rebased from #1187 , merge after that one**

### CODE REVIEW

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!builld

**Comment by szalpal** [comment]
!build

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 1]
```diff
@@ -39,67 +40,28 @@ struct SampleDescriptor {
...
- * Coverts Roi to flattened TensorListShape
- *
- * Flattened TensorListShape, is a TensorListShape, which doesn't have `channels`
- * as separate dimension. Instead, Width dimension is larger by `nchannels` number of times.
- * This is because, the analysis is channel-agnostic.
- *
- * Function assumes, that memory layout is *HWC, and the Roi
- * is represented as [[x_lo, y_lo], [x_hi, y_hi]].
- * Therefore, while copying, order of values needs to be reversed.
+ * TODO
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: missing documentation

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 2]
```diff
@@ -39,67 +40,28 @@ struct SampleDescriptor {
...
-    TensorShape<ndims> ts;
-    auto e = roi.extent();
-    auto ridx = ndims;
-    ts[--ridx] = e[0] * nchannels;
-    for (size_t idx = 1; idx < ndims; idx++) {
-      ts[--ridx] = e[idx];
-    }
-    ret.set_tensor_shape(i++, ts);
+TensorShape<ndims - 1> Flatten(const TensorShape<ndims> &shape) {
+  TensorShape<ndims - 1> ret;
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: static_assert ndims >= 2

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 3]
```diff
@@ -39,67 +40,28 @@ struct SampleDescriptor {
...
-    for (int i = 0; i < shapes.num_samples(); i++) {
-      ret[i] = whole_image(shapes[i]);
-    }
-  } else {
-    for (size_t i = 0; i < rois.size(); i++) {
-      ret[i] = intersection(rois[i], whole_image(shapes[i]));
-    }
+template <size_t ndims>
+TensorListShape<ndims - 1> Flatten(const TensorListShape<ndims> &shape) {
+  TensorListShape<ndims - 1> ret(shape.size());
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: static assert ndims >= 2

File: dali/kernels/imgproc/roi.h [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ */
+template <size_t ndims>
+using Roi = Box<ndims, int>;
+
+namespace detail {
+
+/**
+ * Create Roi for whole image defined by TensorShape
+ */
+template <size_t spatial_dims>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi.h]: maybe more intuitive to have ndims and then `spatial_dims = ndims - 1` in the function body?

File: dali/kernels/imgproc/roi.h [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    ret[ridx--] = e[idx];
+  }
+  return ret;
+}
+
+
+/**
+ * Convenient overload for batch processing (creating TensorListShape)
+ */
+template <size_t ndims>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi.h]: ndims should total number of dimensions. If you need spatial dims define it in the body

File: dali/kernels/imgproc/roi.h [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+ * If `rois` is not provided (roi == nullptr), that means whole image is analysed:
+ * return a Roi, that has the same size as the input image
+ *
+ * If `rois` is provided, adjusted Roi is an intersection of provided Roi and the image.
+ *
+ * Assumes HWC memory layout
+ */
+template <size_t spatial_dims>
+Roi<spatial_dims>
+AdjustRoi(const Roi<spatial_dims> *roi, const TensorShape<spatial_dims + 1> &shape) {
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi.h]: I'd stick to same template argument `ndims` as total number of dimensions and you can define `spatial_dims` as `ndims - 1` when necessary

File: dali/kernels/imgproc/roi.h [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+template <size_t ndims>
+using Roi = Box<ndims, int>;
+
+namespace detail {
+
+/**
+ * Create Roi for whole image defined by TensorShape
+ */
+template <size_t spatial_dims>
+Roi<spatial_dims> whole_image(const TensorShape<spatial_dims + 1> &shape) {
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi.h]:
```suggestion
Roi<spatial_dims> WholeImage(const TensorShape<spatial_dims + 1> &shape) {
```
to match naming convention in this file

File: dali/kernels/imgproc/roi.h [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * 1. If `rois` is empty, that means whole image is analysed: return a batch of Rois, where
+ *    every Roi has the same size as the input image
+ * 2. If `rois` is not empty, it is assumed, that Roi is provided for every image in batch.
+ *    In this case, final Roi is an intersection of provided Roi and the image.
+ *    (This is a sanity-check for Rois, that can be larger than image)
+ *
+ * Assumes HWC memory layout
+ */
+template <size_t spatial_dims>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi.h]: make it ndims

File: dali/kernels/imgproc/roi_test.cc [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+           {5, 6}};
+  Roi roi3{{0, 0},
+           {0, 0}};
+  Rois rois = {roi1, roi2, roi3};
+
+  std::vector<TensorShape<-1>> ref = {{3, 3, nchannels},
+                                      {4, 5, nchannels},
+                                      {0, 0, nchannels}};
+  auto shs = ShapeFromRoi(rois, nchannels);
+  EXPECT_TRUE(shs == TensorListShape<-1>(ref));
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi_test.cc]: EXPECT_EQ

File: dali/kernels/imgproc/roi_test.cc [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    auto sh = ::dali::kernels::ShapeFromRoi(roi, 666);
+    TensorShape<3> ref_sh = {4, 5, 666};
+    ASSERT_EQ(ref_sh, sh);
+  }
+  {
+    Roi roi{{0, 0},
+            {0, 0}};
+    auto sh = ::dali::kernels::ShapeFromRoi(roi, 666);
+    TensorShape<3> ref_sh = {0, 0, 666};
+    ASSERT_EQ(ref_sh, sh);
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi_test.cc]: EXPECT_EQ (same above)

File: dali/kernels/imgproc/roi_test.cc [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+          {{5, 6}, {7, 8}},
+          {0,      1}
+  };
+  auto res = AdjustRoi(rois, tls);
+  ASSERT_EQ(ref.size(), res.size());
+  for (size_t i = 0; i < ref.size(); i++) {
+    EXPECT_EQ(ref[i], res[i]);
+  }
+}
+
```
Comment (jantonguirao) [inline dali/kernels/imgproc/roi_test.cc]: add a test case where the Roi is exactly at the edge of the image

File: dali/kernels/imgproc/roi_test.cc (hunk 10 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi_test.cc]: Done

File: dali/kernels/imgproc/roi_test.cc (hunk 9 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi_test.cc]: Done

File: dali/kernels/imgproc/roi.h (hunk 8 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Dpne

File: dali/kernels/imgproc/roi.h (hunk 7 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Done

File: dali/kernels/imgproc/roi.h (hunk 6 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Done

File: dali/kernels/imgproc/roi.h (hunk 5 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Done

File: dali/kernels/imgproc/roi.h (hunk 4 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Done

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h (hunk 3 above)
Comment (szalpal) [reply dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: Done

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h (hunk 2 above)
Comment (szalpal) [reply dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: Done

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h (hunk 1 above)
Comment (szalpal) [reply dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: Done

File: dali/kernels/imgproc/roi_test.cc (hunk 11 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi_test.cc]: Doone

File: dali/kernels/imgproc/color_manipulation/brightness_contrast.h [hunk 12]
```diff
@@ -22,33 +22,12 @@
...
-  for (size_t idx = 0; idx < ndims; idx++) {
-    ret[ridx--] = e[idx];
-  }
-  return ret;
-}
-
-}  // namespace brightness_contrast
-
-
 template<typename InputType, typename OutputType, size_t ndims = 3>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast.h]: shouldn't it be `int ndims`?

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 13]
```diff
@@ -39,67 +40,40 @@ struct SampleDescriptor {
...
- * Function assumes, that memory layout is *HWC, and the Roi
- * is represented as [[x_lo, y_lo], [x_hi, y_hi]].
- * Therefore, while copying, order of values needs to be reversed.
+ * E.g. [640, 480, 3] -> [640, 1440]
+ *
+ * The reason is that BrightnessContrast calculations are channel-agnostic
+ * (the same operation is applied for every channel), therefore BlockSetup
+ * and SampleDescriptor don't need to know about channels.
  */
 template <size_t ndims>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: int ndims

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 14]
```diff
@@ -39,67 +40,40 @@ struct SampleDescriptor {
...
-
-  if (rois.empty()) {
-    for (int i = 0; i < shapes.num_samples(); i++) {
-      ret[i] = whole_image(shapes[i]);
-    }
-  } else {
-    for (size_t i = 0; i < rois.size(); i++) {
-      ret[i] = intersection(rois[i], whole_image(shapes[i]));
-    }
+template <size_t ndims>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: int ndims

File: dali/kernels/imgproc/roi_test.cc (hunk 10 above)
Comment (jantonguirao) [reply dali/kernels/imgproc/roi_test.cc]: not done

File: dali/kernels/imgproc/color_manipulation/brightness_contrast.h (hunk 12 above)
Comment (jantonguirao) [reply dali/kernels/imgproc/color_manipulation/brightness_contrast.h]: As agreed, you'll change that in next PR

**Comment by szalpal** [comment]
!build

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 15]
```diff
@@ -39,67 +40,40 @@ struct SampleDescriptor {
...
-    assert(all_coords(roi.hi >= roi.lo) && "Cannot create a tensor shape from an invalid Box");
-    TensorShape<ndims> ts;
-    auto e = roi.extent();
-    auto ridx = ndims;
-    ts[--ridx] = e[0] * nchannels;
-    for (size_t idx = 1; idx < ndims; idx++) {
-      ts[--ridx] = e[idx];
-    }
-    ret.set_tensor_shape(i++, ts);
+TensorShape<ndims - 1> Flatten(const TensorShape<ndims> &shape) {
```
Comment (klecki) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]:
```suggestion
TensorShape<ndims - 1> FlattenChannels(const TensorShape<ndims> &shape) {
```
or
```suggestion
TensorShape<ndims - 1> FlattenLast(const TensorShape<ndims> &shape) {
```

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h [hunk 16]
```diff
@@ -39,67 +40,40 @@ struct SampleDescriptor {
...
+  static_assert(ndims >= 2, "If there are less than 2 dims, there's nothing to flatten...");
```
Comment (klecki) [inline dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: Doesn't look formal enough
```suggestion
  static_assert(ndims >= 2, "If there are less than 2 dims, there's nothing to flatten.");
```

File: dali/kernels/imgproc/roi.h [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+ * Additionally, by definition, ROI is top-left inclusive and bottom-right exclusive.
+ * That means, that `Roi.lo` point is included in actual ROI and `Roi.hi` point isn't.
+ */
+template <size_t ndims>
+using Roi = Box<ndims, int>;
+
+namespace detail {
+
+/**
+ * Create Roi for whole image defined by TensorShape
```
Comment (klecki) [inline dali/kernels/imgproc/roi.h]:
```suggestion
 * Create a Roi with size matching the whole image
```

**Comment by szalpal** [comment]
!build

File: dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h (hunk 15 above)
Comment (szalpal) [reply dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h]: Done

File: dali/kernels/imgproc/roi.h (hunk 17 above)
Comment (szalpal) [reply dali/kernels/imgproc/roi.h]: Done
